src/main/kotlin/tanvd/grazi/ide/ui/GraziRulesTree.kt

```diff
@@ -0,0 +1,279 @@
+import com.intellij.ui.*
+import com.intellij.ui.SimpleTextAttributes.REGULAR_ATTRIBUTES
+import com.intellij.ui.SimpleTextAttributes.REGULAR_BOLD_ATTRIBUTES
+import com.intellij.util.ui.JBUI
+import com.intellij.util.ui.UIUtil
+import com.intellij.util.ui.tree.TreeUtil
+import org.languagetool.rules.Category
+import org.languagetool.rules.Rule
+import org.picocontainer.Disposable
+import tanvd.grazi.GraziConfig
+import tanvd.grazi.language.Lang
+import tanvd.grazi.language.LangTool
+import java.awt.BorderLayout
+import java.util.*
+import javax.swing.JPanel
+import javax.swing.JTree
+import javax.swing.SwingUtilities
+import javax.swing.tree.DefaultMutableTreeNode
+import javax.swing.tree.DefaultTreeModel
+import kotlin.Comparator
+import kotlin.collections.HashSet
+
+private data class RuleWithLang(val rule: Rule, val lang: Lang, val enabled: Boolean, var state: Boolean) {
+    val category: Category
+        get() = rule.category
+}
+
+private typealias RulesMap = SortedMap<Lang, SortedMap<Category, MutableList<RuleWithLang>>>
+
+private fun LangTool.allRulesWithLangs(): RulesMap {
+    val result = TreeMap<Lang, SortedMap<Category, MutableList<RuleWithLang>>>()
+    GraziConfig.state.enabledLanguages.forEach { lang ->
+        val categories = TreeMap<Category, MutableList<RuleWithLang>>(Comparator<Category> { o1, o2 -> o1!!.name.compareTo(o2!!.name) })
+        val activeRules = this[lang].allActiveRules.toHashSet()
+        val usedRules = HashSet<String>() // prevent subrules with same ids
+
+        this[lang].allRules.forEach {rule ->
+            if (rule.id !in usedRules) {
+                usedRules.add(rule.id)
+                if (rule in activeRules) {
+                    categories.getOrPut(rule.category, ::LinkedList).add(RuleWithLang(rule, lang, enabled = true, state = true))
+                } else {
+                    categories.getOrPut(rule.category, ::LinkedList).add(RuleWithLang(rule, lang, enabled = false, state = false))
+                }
+            }
+        }
+
+        if (categories.isNotEmpty()) result[lang] = categories
+    }
+
+    return result
+}
+
+class GraziRulesTree(selectionListener: (meta: Any) -> Unit) : Disposable {
+    private val state = HashMap<String, RuleWithLang>()
+    val panel = JPanel(BorderLayout())
+
+    private val tree: CheckboxTree = CheckboxTree(object : CheckboxTree.CheckboxTreeCellRenderer(true) {
+
+        override fun customizeRenderer(tree: JTree?, node: Any?, selected: Boolean, expanded: Boolean, leaf: Boolean, row: Int, hasFocus: Boolean) {
+            if (node !is CheckedTreeNode) return
+
+            val background = UIUtil.getTreeBackground(selected, true)
+            val text = getNodeText(node)
+            val attrs = if (node.userObject is RuleWithLang) REGULAR_ATTRIBUTES else REGULAR_BOLD_ATTRIBUTES
+            UIUtil.changeBackGround(this, background)
+
+            SearchUtil.appendFragments(
+                    filter?.filter, // it can be null
```

Comment:
How it can be `null`? It is declared right under this code.

Reply:
Unfortunately it can :(